File: brainfuck.lua
local module = {}

module.BFInterpreter = {}

function module.BFInterpreter:new(source, cells, input, output)
    local instance = {}

    setmetatable(instance, self)

    self.__index = self

    if input == nil then
        function input()
            while true do
                local result = string.byte(io.read(1))
                if math.type(result) == "integer" and result >= 0 and result <= 255 then
                    return result
                end
            end
        end
    end

    if output == nil then
        function output(byte)
            io.write(string.char(byte))
            io.flush()
        end
    end

    if type(source) ~= "string" then
        error("BFInterpreter() source must be string")
    end
    if math.type(cells) ~= "integer" and cells ~= nil then
        error("BFInterpreter() cells must be integer")
    end
    if type(input) ~= "function" then
        error("BFInterpreter() input must be a function")
    end
    if type(output) ~= "function" then
        error("BFInterpreter() output must be a function")
    end

    if cells ~= nil and cells <= 0 then
        error("BFInterpreter() cells must be greater than 0")
    end

    self.running = false

    self.source = source
    self.cells = cells
    self.input = input
    self.output = output

    self._tokenIndex = 0
    self._tokens = {}
    self._bracketMap = {}

    local stack = {}
    local comment = false
    local tokenIndex = 1

    for position = 1, #source do
        local character = source:sub(position, position)

        if not comment and character == '#' then
            comment = true
        elseif comment and character == '\n' then
            comment = false
        end

        if not comment and (
            character == '<' or
            character == '>' or
            character == '+' or
            character == '-' or
            character == '.' or
            character == ',' or
            character == '[' or
            character == ']'
        ) then

            if character == '[' then
                table.insert(stack, tokenIndex)

            elseif character == ']' then
                if #stack == 0 then
                    error("unbalanced brackets")
                end

                local startIndex = table.remove(stack)

                self._bracketMap[startIndex] = tokenIndex
                self._bracketMap[tokenIndex] = startIndex
            end

            table.insert(self._tokens, {position, character})
            tokenIndex = tokenIndex + 1
        end
    end

    if #stack ~= 0 then
        error("unbalanced brackets")
    end

    return instance
end

function module.BFInterpreter:start()
    if self.running then
        return
    end

    self._tokenIndex = 0

    self.running = true
    self.memory = {}
    self.point = 1

    if self.cells ~= nil then
        for i = 1, self.cells do
            self.memory[i] = 0
        end
    end
end

function module.BFInterpreter:step()
    if not self.running then
        return
    end

    self._tokenIndex = self._tokenIndex + 1

    if self._tokenIndex > #self._tokens then
        self.running = false
        return
    end

    local position, character = table.unpack(self._tokens[self._tokenIndex])
    local dataPointer = self.memory[self.point] or 0

    if character == '>' then
        self.point = self.point + 1
        if self.cells == nil then
            if self.point > #self.memory then
                self.memory[self.point] = 0
            end
        elseif self.point > self.cells then
            error("pointer out of bounds")
        end

    elseif character == '<' then
        if self.point == 1 then
            error("pointer out of bounds")
        end
        self.point = self.point - 1

    elseif character == '+' then
        self.memory[self.point] = (dataPointer + 1) % 256

    elseif character == '-' then
        self.memory[self.point] = (dataPointer - 1) % 256

    elseif character == ',' then
        local input = self.input()

        if not (math.type(input) == "integer" and input >= 0 and input <= 255) then
            error("BFInterpreter() input must be returns unsigned 8-bit integer")
        end

        self.memory[self.point] = input

    elseif character == '.' then
        self.output(dataPointer)

    elseif character == '[' and dataPointer == 0 then
        self._tokenIndex = self._bracketMap[self._tokenIndex]

    elseif character == ']' and dataPointer ~= 0 then
        self._tokenIndex = self._bracketMap[self._tokenIndex]
    end

    return {self.point, position, character}
end

function module.BFInterpreter:stop(cleanUp)
    if not self.running then
        return
    end

    self.running = false

    if cleanUp == nil or cleanUp then
        self.memory = nil
        self.point = nil
    end
end

function module.BFExec(source, cells, input, output)
    local interpreter = module.BFInterpreter:new(source, cells, input, output)

    interpreter:start()

    repeat interpreter:step() until not interpreter.running

    interpreter:stop()
end

return module

File: brainfuck.py
import sys

if sys.platform == 'win32':
    from msvcrt import getch
else:
    import termios
    import tty

    def getch():
        fileDescriptor = sys.stdin.fileno()
        oldSettings = termios.tcgetattr(fileDescriptor)

        try:
            tty.setraw(sys.stdin.fileno())
            character = sys.stdin.read(1).encode()
        finally:
            termios.tcsetattr(fileDescriptor, termios.TCSADRAIN, oldSettings)

        return character

__all__ = [
    'BFInterpreter',
    'BFExec'
]

class BFInterpreter:

    def __init__(self, source, *, cells=None, input=None, output=None):
        if input is None:
            def input():
                while True:
                    result = ord(getch())
                    if 0 <= result <= 255:
                        return result

        if output is None:
            def output(byte):
                sys.stdout.write(chr(byte))
                sys.stdout.flush()

        if not isinstance(source, str):
            raise TypeError("BFInterpreter() source must be string")
        if not isinstance(cells, (int, type(None))):
            raise TypeError("BFInterpreter() cells must be integer")
        if not callable(input):
            raise TypeError("BFInterpreter() input must be callable")
        if not callable(output):
            raise TypeError("BFInterpreter() output must be callable")

        if cells is not None and cells <= 0:
            raise ValueError("BFInterpreter() cells must be greater than 0")

        self.running = False

        self.source = source
        self.cells = cells
        self.input = input
        self.output = output

        self._tokenIndex = -1
        self._tokens = []
        self._bracketMap = {}

        stack = []
        comment = False
        tokenIndex = 0

        for position, character in enumerate(source):

            if not comment and character == '#':
                comment = True
            elif comment and character == '\n':
                comment = False

            if not comment and character in '<>+-.,[]':

                if character == '[':
                    stack.append(tokenIndex)

                elif character == ']':
                    if not stack:
                        raise SyntaxError("unbalanced brackets")

                    startIndex = stack.pop()

                    self._bracketMap[startIndex] = tokenIndex
                    self._bracketMap[tokenIndex] = startIndex

                self._tokens.append((position, character))
                tokenIndex += 1

        if stack:
            raise SyntaxError("unbalanced brackets")

    def start(self):
        if self.running:
            return

        self._tokenIndex = -1

        self.running = True
        self.memory = [0]
        self.point = 0

        if self.cells is not None:
            self.memory *= self.cells

    def step(self):
        if not self.running:
            return

        self._tokenIndex += 1

        if self._tokenIndex >= len(self._tokens):
            self.running = False
            return

        position, character = self._tokens[self._tokenIndex]
        dataPointer = self.memory[self.point]

        if character == '>':
            self.point += 1
            if self.cells is None:
                if self.point == len(self.memory):
                    self.memory.append(0)
            elif self.point >= self.cells:
                raise IndexError("pointer out of bounds")

        elif character == '<':
            if self.point == 0:
                raise IndexError("pointer out of bounds")
            self.point -= 1

        elif character == '+':
            self.memory[self.point] = (dataPointer + 1) % 256

        elif character == '-':
            self.memory[self.point] = (dataPointer - 1) % 256

        elif character == ',':
            input = self.input()

            if not (isinstance(input, int) and 0 <= input <= 255):
                raise TypeError("BFInterpreter() input must be returns unsigned 8-bit integer")

            self.memory[self.point] = input

        elif character == '.':
            self.output(dataPointer)

        elif character == '[' and dataPointer == 0:
            self._tokenIndex = self._bracketMap[self._tokenIndex]

        elif character == ']' and dataPointer != 0:
            self._tokenIndex = self._bracketMap[self._tokenIndex]

        return self.point, position, character

    def stop(self, cleanUp=True):
        if not self.running:
            return

        self.running = False

        if cleanUp:
            del self.memory, self.point

    def __iter__(self):
        self.start()

        return self

    def __next__(self):
        result = self.step()

        if result is None:
            raise StopIteration

        return result

def BFExec(source, **kwargs):
    interpreter = BFInterpreter(source, **kwargs)

    for _ in interpreter:
        pass

    interpreter.stop()